src/library/scala/collection/immutable/Vector.scala

```diff
@@ -30,11 +30,21 @@ object Vector extends StrictOptimizedSeqFactory[Vector] {
 
   def from[E](it: collection.IterableOnce[E]): Vector[E] =
     it match {
-      case as: ArraySeq[E] if as.length <= 32 && as.unsafeArray.isInstanceOf[Array[AnyRef]] =>
+      case as: ArraySeq[E] if as.length <= 32 =>
         if (as.isEmpty) NIL
         else {
           val v = new Vector(0, as.length, 0)
-          v.display0 = as.unsafeArray.asInstanceOf[Array[AnyRef]]
+          v.display0 = as.unsafeArray match {
+            case array: Array[AnyRef] => array
+            case array =>
+              val display0 = new Array[AnyRef](array.length)
```

Comment:
probably simpler to call `toArray` rather than manually copying

```scala
case _ => as.toArray[AnyRef]
```

Reply:
sadly that won't compile because we don't have that `E <: AnyRef`, so `def toArray[B >: E]` won't work.

Reply:
`as.toArray[Any]`.asInstanceOf[Array[AnyRef]]`

Reply:
Hmm looked into that, seems it goes through a lot many intermediate method calls, with additional bounds checks and math checks, to eventually just do this exact thing, so I reckon this is faster but I could try

Reply:
you might also be able to do `Array.from[AnyRef]` or `Array.from[Any]`

Reply:
That code path would make two additional allocs: 
* the implicitly created wrapper around the `Array` that allows it to conform to `IterableOnce`
* the wrapper's `.iterator`

Reply:
`Array.copyAs` is the right method for this task. Maybe the manually inlined version here is a bit faster.